### PR TITLE
RBF: add linear polynomial term in RBF interpolation

### PR DESCRIPTION
**Pull request by edoardolombardi**
The pull introduces a regularization linear multivariate polynomial to the RBF basis during RBF interpolation.

The polynomial is based on the btipit class ReconstructionPolynomial and specialized only for a first degree polynomial. 

Configurations of RBF nodes leading to undetermined linear systems, because identical rows/columns provided by polynomial basis terms, are automatically managed by reduction of the variables of the multivariate polynomial. 

A custom class derived by RBFKernel has now to override new pure virtual methods aimed to the initialization of the polynomial, the identification of the independent variables and the evaluation of the polynomial basis.

A new integration test is added: test_RBF_00004.

### CODE REVIEW

**Review by roccoarpa — APPROVED** [review]
Just optional, if you can, please improve the test 00004 considering variable support radii instead of a homogeneous one.

File: test/integration_tests/RBF/test_RBF_00004.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	zDispl.resize(nCP);
+    zDispl[0] = 1.;
+    zDispl[1] = 1.;
+    zDispl[2] = 2.;
+    zDispl[3] = 1.;
+
+    
+     //RBF used to directly interpolate zDispl field
+    bitpit::RBF   paraMorph;
+	RBFBasisFunction funct = RBFBasisFunction::WENDLANDC2;
```
Comment (kgkara) [inline test/integration_tests/RBF/test_RBF_00004.cpp:69]: @edoardolombardi just a question: I am not sure, so I ask here, isn't the usage of the polynomials more reasonable only when you make use of thin splines (see http://www.scholarpedia.org/article/Radial_basis_function#Further_example_with_linear_additional_terms in the section of polynomias)? If you agree, would you add a thin spline to bitpit, of the form:
f(r)= r ^2 * log(r)? 
If you decide to do it within this PR, maybe line 69 can be substituted?

**Review by kgkara — CHANGES_REQUESTED** [review]

File: src/RBF/rbf.hpp [hunk 2]
```diff
@@ -64,6 +66,28 @@ enum class RBFMode {
 };
 
 class RBFKernel{
+    /*!
+     * Linear Polynomial class. A linear multivariate polynomial in function of the 
+     * variables defining the RBF nodes is used to regularize the interpolation 
+     * problem through RBFs. Used only when RBFMode::PARAM is enabled.
```
[the remark this answers is not in the document]
Comment (edoardolombardi) [reply src/RBF/rbf.hpp]: fixed in https://github.com/optimad/bitpit/pull/434/commits/e179283a77b6f9e26062e443ffa82f4917a6fd55

File: test/integration_tests/RBF/test_RBF_00004.cpp (hunk 1 above)
Comment (edoardolombardi) [reply test/integration_tests/RBF/test_RBF_00004.cpp]: Added in https://github.com/optimad/bitpit/pull/434/commits/b5c33c96b33e7eeb4fd8111b44a039515cb199e5 and https://github.com/optimad/bitpit/pull/434/commits/e8a8c600ffaaec28629801f1b529cec9165594c9.

File: test/integration_tests/RBF/test_RBF_00003.cpp [hunk 3]
```diff
@@ -384,7 +384,7 @@ int subtest_001()
 			controlNodes[i][2] = lims[2][1];
 		}
 		
-		double maxDim = 0.0;
+        double maxDim = 0.0;
```
Comment (andrea-iob) [inline test/integration_tests/RBF/test_RBF_00003.cpp:387]: Unrelated.

File: test/integration_tests/RBF/test_RBF_00003.cpp [hunk 4]
```diff
@@ -419,8 +419,8 @@ int subtest_001()
 			RBFBasisFunction funct = RBFBasisFunction::WENDLANDC2;
 			
 			paraMorph.setFunction(funct);
-			
-			double maxVal = 0.0;
+
+            double maxVal = 0.0;
 			maxval(zDispl, maxVal);
```
Comment (andrea-iob) [inline test/integration_tests/RBF/test_RBF_00003.cpp:424]: Unrelated.

File: src/RBF/rbf.cpp [hunk 5]
```diff
@@ -62,8 +64,8 @@ namespace bitpit {
 RBFKernel::RBFKernel()
     : m_supportRadii(1, 1.)
 {
-    m_nodes         = 0;
-    m_fields        = 0;
+    m_nodes  = 0;
+    m_fields = 0;
```
Comment (andrea-iob) [inline src/RBF/rbf.cpp:68]: Unrelated.

File: src/RBF/rbf.cpp [hunk 6]
```diff
@@ -72,18 +74,22 @@ RBFKernel::RBFKernel()
     m_weight.clear();
     m_activeNodes.clear();
 
-    setFunction( RBFBasisFunction::WENDLANDC2);
+    setFunction(RBFBasisFunction::WENDLANDC2);
```
Comment (andrea-iob) [inline src/RBF/rbf.cpp:77]: Unrelated.

File: src/RBF/rbf.cpp [hunk 7]
```diff
@@ -72,18 +74,22 @@ RBFKernel::RBFKernel()
...
+
+    m_polyActiveBasis.clear();
+    m_polynomial.clear();
 }
 
 /*!
  * Copy Constructor
  */
-RBFKernel::RBFKernel(const RBFKernel & other)
+RBFKernel::RBFKernel(const RBFKernel &other)
```
Comment (andrea-iob) [inline src/RBF/rbf.cpp:87]: Unrelated.

File: src/RBF/rbf.hpp [hunk 8]
```diff
@@ -64,6 +66,28 @@ enum class RBFMode {
...
+     */
+    class LinearPolynomial : public ReconstructionPolynomial
+    {
+private:
+        int m_dim;                          /**< Polynomial variables space dimension */
+        int m_fields;                       /**< Number of equations of the polynomial object */
+
+    public:
+        LinearPolynomial();
+        void setDefault();
```
Comment (andrea-iob) [inline src/RBF/rbf.hpp:85]: reset seems a better name.

**Comment by andrea-iob** [comment]
You have mixed tons of cosmetic changes with the functional changes associated with the pull request. It's hard for me to understand what are the real changes. I'll leave the review to @roccoarpa and @kgkara.

**Comment by andrea-iob** [comment]
Please squash together all the commits before merging.

**Comment by edoardolombardi** [comment]
> You have mixed tons of cosmetic changes with the functional changes associated with the pull request. It's hard for me to understand what are the real changes. I'll leave the review to @roccoarpa and @kgkara.

Ah, yes, sorry for that. It was clang-format enabled for modified code. If you want I can try to isolate format changes with implementations.

**Comment by edoardolombardi** [comment]
The new forced pushed branch is cleaned of formatting modifications. 
Let me know if you'd like to add your review, @andrea-iob, thanks.

File: src/RBF/rbf.hpp [hunk 9]
```diff
@@ -64,6 +66,28 @@ enum class RBFMode {
...
+        void clear();
```
Comment (andrea-iob) [inline src/RBF/rbf.hpp:86]: What's the point of the setDefault/reset method? If dimension and field count are set to zero, the polynomial will be no longer valid. I don't see why setDefault/reset should be called instead of calling clear. If, for performance reasons, you don't want to release the memory  hold by the polynomial, you can add a release argument to the clear method and pass that argument when you clear the polynomial. I would delete the method setDefault/reset.

File: src/RBF/rbf.cpp [hunk 10]
```diff
@@ -1692,13 +1697,27 @@ double rbf::c2c2( double dist )
...
-            return 0.;
-        } else{
-            return 0.5 + 0.5 * std::cos(dist*BITPIT_PI);
-        }
+double rbf::cosinus( double dist )
+{
+    if( dist > 1) {
+        return 0.;
+    } else{
+        return 0.5 + 0.5 * std::cos(dist*BITPIT_PI);
```
Comment (andrea-iob) [inline src/RBF/rbf.cpp:1705]: This seem just an unrelated cosmetic fix.

File: src/RBF/rbf.hpp [hunk 11]
```diff
@@ -53,6 +53,7 @@ enum class RBFBasisFunction {
     C1C2       = 12, /**< Compact biquadratic funct, C1 on r=0, C2 on r=1, 0 outside */
     C2C2       = 13, /**< Compact poly (degree 5) funct, C2 on r=0, C2 on r=1, 0 outside */
     COSINUS    = 14, /**< Compact cosinusoidal funct, value of 1 on r=0, 0 outside */
+    THINPLATE  = 15, /**< Non compact thin plate funct */
```
Comment (andrea-iob) [inline src/RBF/rbf.hpp:56]: "Non-compact"

File: src/RBF/rbf.cpp [hunk 12]
```diff
@@ -1181,7 +1381,105 @@ double RBF::calcDist(int i, int j)
...
+    std::array<double, 3> coord(m_node[0]);
+
+    // Check if at least one node has one of the coordinates
+    // different from the reference ones and enable the related
+    // polynomial term
+    std::set<int> coordinatesToCheck({0, 1, 2});
+    double tol = 1.0e-12;
+    for (const auto &i : activeSet) {
+        std::array<double, 3> point = m_node[i];
+        for (auto it = coordinatesToCheck.begin(); it != coordinatesToCheck.end();) {
```
Comment (andrea-iob) [inline src/RBF/rbf.cpp:1434]: It seems a quite convoluted way to do a loop over 0, 1, and 2. Maybe something like the following may be easy to understand:
```
for (int d = 0; d < 2; ++d) {
    if (!utils::DoubleFloatingEqual()(coord[d], point[d], tol, tol)) {
        m_polyActiveBasis.insert(d+1);
    }
}
```

File: src/RBF/rbf.cpp [hunk 13]
```diff
@@ -1181,7 +1381,105 @@ double RBF::calcDist(int i, int j)
...
+ * the related coordinate if disable in the polynomial function (no dependency 
+ * can be found).
+ */
+void RBF::initializePolynomialActiveBasis()
+{
+    std::vector<int> activeSet(RBFKernel::getActiveSet());
+
+    // Initialize active terms for an only constant linear polynomial
+    m_polyActiveBasis.clear();
+    m_polyActiveBasis.insert(0);
```
Comment (andrea-iob) [inline src/RBF/rbf.cpp:1423]: The documentation says "Active terms of linear polynomial, -1 is constant, 0,1,2 the system coordinates", it seems to me that the documentation should be updated.

**Review by andrea-iob — APPROVED** [review]

File: src/RBF/rbf.hpp [hunk 14]
```diff
@@ -74,12 +98,15 @@ class RBFKernel{
...
     std::vector<double>                 m_error;    /**<Interpolation error of a field evaluated on each RBF node (auxiliary memeber used in Greedy algorithm).*/
 
-    protected:
+protected:
     std::vector<std::vector<double>>    m_value;    /**< displ value to be interpolated on RBF nodes */
     std::vector<std::vector<double>>    m_weight;   /**< weight of your RBF interpolation */
     std::vector<bool>                   m_activeNodes;   /**<Vector of active/inactive node (m_activeNodes[i] = true/false -> the i-th node is used/not used during RBF evaluation).*/   
     int m_maxFields;                                /**< fix the maximum number of fields that can be added to your class*/
     int m_nodes;                                    /**<Number of RBF nodes.*/
+    std::set<int> m_polyActiveBasis;                /**< Active terms of linear polynomial, -1 is constant, 0,1,2 the system coordinates */
```
[the remark this answers is not in the document]
Comment (edoardolombardi) [reply src/RBF/rbf.hpp:110]: In case of skipping checks on coordinates already set as in the current implementation yes, it can be a vector. 
In case of loop always from 0 to 2 as proposed in https://github.com/optimad/bitpit/pull/434#discussion_r1417048034, I think that has to remain a set to avoid multiple insert of the same coordinate index.

File: src/RBF/rbf.cpp [hunk 15]
```diff
@@ -1181,7 +1381,105 @@ double RBF::calcDist(int i, int j)
...
+
+    // Initialize reference coordinates with the first node
+    std::array<double, 3> coord(m_node[0]);
+
+    // Check if at least one node has one of the coordinates
+    // different from the reference ones and enable the related
+    // polynomial term
+    std::set<int> coordinatesToCheck({0, 1, 2});
+    double tol = 1.0e-12;
```
[the remark this answers is not in the document]
Comment (edoardolombardi) [reply src/RBF/rbf.cpp]: In the next commit I removed the tol definition and left the static bitpit relative and absolute tolerances used as default values by DoubleFloatingEqual operator.
